src/xpk/core/ray.py

```diff
@@ -0,0 +1,198 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import re
+from ..utils import write_tmp_file, xpk_print, xpk_exit
+from .commands import run_command_with_updates_retry, run_command_for_value
+
+
+ray_cluster_crd_yaml = """apiVersion: v1
+kind: Namespace
+metadata:
+  name: ray
+---
+apiVersion: ray.io/v1
+kind: RayCluster
+metadata:
+  name: raycluster
+  namespace: ray
+spec:
+  rayVersion: '{version}'
+  headGroupSpec:
+    rayStartParams: {{}}
+    #pod template
+    template:
+      spec:
+        containers:
+        - name: ray-head
+          image: rayproject/ray:{version}
+          resources:
+            limits:
+              cpu: {head_cpu}
+              memory: {head_mem}
+            requests:
+              cpu: {head_cpu}
+              memory: {head_mem}
+          ports:
+          - containerPort: 6379
+            name: gcs-server
+          - containerPort: 8265 # Ray dashboard
+            name: dashboard
+          - containerPort: 10001
+            name: client
+          - containerPort: 8081
+            name: multislice
+  workerGroupSpecs:
+    - replicas: {replicas} # TODO: Set min and max replicas
+      numOfHosts: {num_hosts}
+      minReplicas: {replicas}
+      maxReplicas: {replicas}
+      groupName: workergroup0
+      rayStartParams:
+        block: 'true'
+      template:
+        spec:
+          containers:
+            - name: ray-worker
+              image: rayproject/ray:{version}
+              resources:
+                limits:
+                  cpu: {worker_cpu}
+                  google.com/tpu: {chips_per_vm}
+                  memory: {worker_mem}
+                requests:
+                  cpu: {worker_cpu}
+                  google.com/tpu: {chips_per_vm}
+                  memory: {worker_mem}
+          nodeSelector:
+            cloud.google.com/gke-tpu-accelerator: {accelerator}
+            cloud.google.com/gke-tpu-topology: {topology}
+"""
+
+
+def install_ray_cluster(args, system) -> int:
+  """Install a RayCluster on the cluster
+
+  Args:
+    args: user provided arguments for running the command.
+    system: system characteristics.
+
+  Returns:
+    0 if successful and 1 otherwise.
+  """
+
+  delete_ray_cluster(args)
+
+  label = 'cloud.google.com/gke-nodepool=default-pool'
+  available_head_cpu, available_head_mem = generate_available_resources(
+      label, args, 0.5
+  )
+
+  label = f'cloud.google.com/gke-tpu-accelerator={system.gke_accelerator}'
+  available_worker_cpu, available_worker_mem = generate_available_resources(
+      label, args, 0.9
+  )
+
+  yml_string = ray_cluster_crd_yaml.format(
+      accelerator=system.gke_accelerator,
+      topology=system.topology,
+      chips_per_vm=system.chips_per_vm,
+      num_hosts=system.vms_per_slice,
+      replicas=args.num_slices,
```

Comment:
maybe this is suppose to the be arg instead of num_nodes?

Reply:
Still works without explicitly adding num_slices as an arg.